File: Android/DepollSoftCommon/src/depollsoft/lib/util/Task.java
package depollsoft.lib.util;

import java.util.ArrayList;
import java.util.List;

public class Task<T>
{
   public static class TaskSource<T>
   {
      private T result;
      private Exception error;
      private boolean isFinished;
      private Task<T> task;

      public Exception getError()
      {
         return this.error;
      }

      public T getResult()
      {
         return this.result;
      }

      public void setError(Exception value)
      {
         this.error = value;
         this.isFinished = true;
         if (this.task != null)
            this.task.finish();
      }

      public void setResult(T value)
      {
         this.result = value;
         this.isFinished = true;
         if (this.task != null)
            this.task.finish();
      }
   }

   private Object continuationsLock = new Object();
   private List<Action<T>> continuations;
   private List<Action<Exception>> errors;
   private TaskSource<T> source;

   public Task(TaskSource<T> source)
   {
      this.continuations = new ArrayList<Action<T>>();
      this.errors = new ArrayList<Action<Exception>>();
      this.source = source;
      this.source.task = this;
      if (source.isFinished)
         this.finish();
   }

   public void continueWith(Action<T> continuation, Action<Exception> error)
   {
      synchronized (this.continuationsLock)
      {
         if (continuation != null)
            this.continuations.add(continuation);
         if (error != null)
            this.errors.add(error);
      }
      if (this.source.isFinished)
         this.finish();
   }

   public void finish()
   {
      synchronized (this.continuationsLock)
      {
         try
         {
            if (this.source.getError() != null)
            {
               for (Action<Exception> error : this.errors)
               {
                  error.invoke(this.source.getError());
               }
               return;
            }
            for (Action<T> continuation : this.continuations)
            {
               continuation.invoke(this.source.getResult());
            }
         }
         finally
         {
            this.continuations.clear();
            this.errors.clear();
         }
      }
   }

   public T waitFor()
   {
      while (!this.source.isFinished);
      return this.source.result;
   }
}

File: DotNet/BarbershopTags/BarbershopTags/Task.cs
using System;

namespace BarbershopTags {
  public class Task<T> {
    private object _ContinuationsLock = new object();
    private Action<T> _Continuations;
    private Action<Exception> _Errors;
    private TaskSource _Source;
    public Task(TaskSource source) {
      _Source = source;
      source.Task = this;
      if (source.IsFinished)
        Finish();
    }

    public void Continue(Action<T> continuation, Action<Exception> error = null) {
      lock (_ContinuationsLock) {
        _Continuations += continuation;
        _Errors += error;
      }
      if (_Source.IsFinished)
        Finish();
    }

    private void Finish() {
      lock (_ContinuationsLock) {
        try {
          if (_Source.Error != null)
            if (_Errors != null) {
              _Errors(_Source.Error);
              return;
            }
          if (_Continuations != null)
            _Continuations(_Source.Result);
        } finally {
          _Continuations = null;
          _Errors = null;
        }
      }
    }

    public class TaskSource {
      private T _Result;
      private Exception _Error;
      internal bool IsFinished { get; private set; }
      internal Task<T> Task { get; set; }
      public T Result {
        get {
          return _Result;
        }
        set {
          _Result = value;
          IsFinished = true;
          if (Task != null)
            Task.Finish();
        }
      }
      public Exception Error {
        get {
          return _Error;
        }
        set {
          _Error = value;
          IsFinished = true;
          if (Task != null)
            Task.Finish();
        }
      }
    }
  }
}
